src/Tasks.UnitTests/DownloadFile_Tests.cs

```diff
@@ -245,6 +245,68 @@ public void RetryOnResponseError()
             _mockEngine.Log.ShouldContain("MSB3924", () => _mockEngine.Log);
         }
 
+        [Fact]
+        public void AbortOnTimeout()
+        {
+            CancellationTokenSource timeout = new CancellationTokenSource();
+            timeout.Cancel();
+            DownloadFile downloadFile = new DownloadFile()
+            {
+                BuildEngine = _mockEngine,
+                HttpMessageHandler = new MockHttpMessageHandler((message, token) =>
+                {
+                    // Http timeouts manifest as "OperationCanceledExceptions" from the handler, simulate that
+                    throw new OperationCanceledException(timeout.Token);
+                }),
+                Retries = 1,
+                RetryDelayMilliseconds = 100,
+                SourceUrl = "http://notfound/foo.txt"
+            };
+
+            downloadFile.Execute().ShouldBeFalse(() => _mockEngine.Log);
+
+            _mockEngine.Log.ShouldContain("MSB3923", () => _mockEngine.Log);
+        }
+
+        [Fact]
+        public async Task NoRunawayLoop()
+        {
+            DownloadFile downloadFile = null;
+            bool failed = false;
+            downloadFile = new DownloadFile()
+            {
+                BuildEngine = _mockEngine,
+                HttpMessageHandler = new MockHttpMessageHandler((message, token) =>
+                {
+                    token.ThrowIfCancellationRequested();
+                    downloadFile.Cancel();
+                    if (!failed)
+                    {
+                        failed = true;
+                        return new HttpResponseMessage(HttpStatusCode.InternalServerError);
+                    }
+
+                    return new HttpResponseMessage(HttpStatusCode.OK)
+                    {
+                        Content = new StringContent("Success!"),
+                        RequestMessage = new HttpRequestMessage(HttpMethod.Get, "http://success/foo.txt")
+                    };
+                }),
+                Retries = 2,
+                RetryDelayMilliseconds = 100,
+                SourceUrl = "http://notfound/foo.txt"
+            };
+
+            var runaway = Task.Run(() => downloadFile.Execute());
+            await Task.Delay(TimeSpan.FromSeconds(1));
+            runaway.IsCompleted.ShouldBeTrue("Task did not cancel");
+            if (!runaway.IsCompleted)
```

Comment:
Isn't this redundant given the previous line? If not, can you add braces around it or put it on one line?

Reply:
I suppose it is, if the assumption is that a failed "Should" will _always_ abort the test.  The problem is that if this isn't trye, the link 306 will wait infinitely.  I wanted to be extra sure the test didn't enter an infinite loop, since it's really hard to report on those.

I'd rather leave the valve, but I'll add the braces and add a comment explaining why it's there.

Reply:
Since xUnit assertions work via exception, I think it's safe to assume that the `ShouldBeTrue` will end the test if it fails. We also have #4973 and https://github.com/dotnet/arcade/pull/3255 coming which should help catch hung tests in the general case.

Reply:
The arcade situation will almost certainly cause it to just... lose the test results, but I'm find removing this test hardening.  Committed an update.